File: apps/worker-notification/sendNotification.ts
import type { NotificationMessage } from "@repo/redis";
import { incrementRetry, isAlreadySent, markAsSentOnce, pushToDlq, shouldThrottle } from "@repo/redis";
import { env } from "./env";
import { channels } from "./channels";
import { getStatusEventType } from "./utils/statusEventType";
import { DEFAULT_RULES } from "./rules";
import { websiteService } from "@repo/shared-repo";
import { logNotification } from "./logger";

const NOTIFICATION_STREAM = env.NOTIFICATION_STREAM;
const MAX_TRIES = 5;

export const sendNotification = async (
    messageId : string,
    msg : NotificationMessage
): Promise<"NO_OP" | "ALREADY_SENT" | "SENT" | "DLQ"> => {

    try{
        
        // 1. Determine status transition
        const statusEventType = getStatusEventType(
            msg.prevStatus,
            msg.currentStatus
        )
        console.log(`Website Status`,statusEventType);
        if(statusEventType === null){
            logNotification({
                websiteId : msg.websiteId,
                status : "NO_OP"
            })
            return "NO_OP";
        }

         // 2. Check mark notification sent ( per message ) / Idempoteny Check
        const alreadySent = await isAlreadySent(
            msg.websiteId,
            statusEventType
        )

        if(alreadySent){
            logNotification({
                websiteId : msg.websiteId,
                eventType : statusEventType,
                status : "ALREADY_SENT"
            })
            return "ALREADY_SENT";
        }
        
        let sentAtLeastOnce = false;
        let userEmail : string | null = null;

        // 3. Apply rules
        for(const rule of DEFAULT_RULES){
            if(!rule.enabled) continue;
            if( rule.notifyOn !== "BOTH" && 
                rule.notifyOn !== statusEventType){
                continue;
            }

           // 4. Rate Limit per notification per channel
           const throttled = await shouldThrottle(
                msg.websiteId,
                rule.channel,
                statusEventType
           )

           if(throttled){
            logNotification({
                websiteId : msg.websiteId,
                channel : rule.channel,
                eventType : statusEventType,
                status : "THROTTLED"
            })
            continue;
           }


            // 5. Send notification

            const channel = channels[rule.channel];
            console.log(`Using channel ${channel} to send notification`);
            if(!channel){
                console.warn(`Channel not registered ${rule.channel}`);
                continue;
            }

            if(!userEmail){
                userEmail = await websiteService.resolveUserEmailForWebsite(
                    msg.websiteId
                );
                console.log(`User email to whom notification will be sent : `,userEmail);
            }
            

            await channels[rule.channel]!.send({
                websiteId : msg.websiteId,
                eventType : statusEventType,
                occurredAt : msg.occurredAt,
                email : userEmail
            })
            sentAtLeastOnce = true;

            logNotification({
                websiteId : msg.websiteId,
                eventType : statusEventType,
                channel : rule.channel,
                status : "SENT"
            })
        }

        if(sentAtLeastOnce){
            await markAsSentOnce(
                msg.websiteId,
                statusEventType
            )
            return "SENT";
        }

       
        return "NO_OP";

    }catch(err){
        
        // 6. Retry Logic
        const retries = await incrementRetry(messageId);
        logNotification({
            messageId,
            websiteId : msg.websiteId,
            status : "FAILED",
            retries : retries.toString(),
            reason : (err as Error).message
        })
        if(retries >= MAX_TRIES){
            await pushToDlq(NOTIFICATION_STREAM,{
                messageId,
                websiteId : msg.websiteId,
                prevStatus : msg.prevStatus,
                currentStatus : msg.currentStatus,
                occurredAt : msg.occurredAt,
                retries : retries.toString(),
                reason: "Max retries exceeded"
            });
            logNotification({
                messageId,
                websiteId : msg.websiteId,
                status : "DLQ",
                retries : retries.toString()
            })
            return "DLQ";
        }

        // exponential backoff
        await new Promise(res => setTimeout(res, retries * 1000));
        throw err;
    }
}

File: apps/api-go/internal/workers/notification-worker/send_notification.go
package notificationworker

import (
	"context"
	"time"

	"github.com/RitikaxG/runState/apps/api-go/internal/channels"
	"github.com/RitikaxG/runState/apps/api-go/internal/domain"
	"github.com/RitikaxG/runState/apps/api-go/internal/service"
)

func (nw *NotificationWorker) SendNotification(
	ctx context.Context,
	msg domain.NotificationMessage,
) error {
	// 1. Determine event type
	eventTypePtr := domain.GetEventType(msg.PrevStatus, msg.CurrentStatus)
	if eventTypePtr == nil {
		LogNotification(map[string]string{
			"websiteId": msg.WebsiteID,
			"status":    string(NotifNoOp),
		})
		return nil
	}

	eventType := *eventTypePtr

	// 2. Idempotency check
	ok, err := nw.redis.MarkAsSentOnce(ctx, msg.WebsiteID, eventType)
	if err != nil {
		return err
	}

	if !ok {
		LogNotification(map[string]string{
			"websiteId": msg.WebsiteID,
			"eventType": string(eventType),
			"status":    string(NotifAlreadySent),
		})
		return nil
	}

	var (
		sentAtLeastOnce bool
		userEmail       string
	)

	// Optional: resolve active incident once
	var incidentID *string
	if nw.incidentRepo != nil {
		activeIncident, err := nw.incidentRepo.GetActiveByWebsiteAndRegion(
			ctx,
			msg.WebsiteID,
			msg.RegionID,
		)
		if err == nil && activeIncident != nil {
			incidentID = &activeIncident.ID
		}
	}

	// 3. Apply rules
	for _, rule := range nw.rules {
		if !ShouldNotify(rule, string(eventType)) {
			continue
		}

		// 4. Throttling
		throttled, err := nw.redis.ShouldThrottle(
			ctx,
			msg.WebsiteID,
			eventType,
			rule.Channel,
		)
		if err != nil {
			return err
		}

		if throttled {
			LogNotification(map[string]string{
				"websiteId": msg.WebsiteID,
				"channel":   rule.Channel,
				"eventType": string(eventType),
				"status":    string(NotifThrottled),
			})
			continue
		}

		// 5. Resolve channel
		channelType := domain.NotificationChannelType(rule.Channel)
		channel, ok := nw.channels[channelType]
		if !ok {
			LogNotification(map[string]string{
				"websiteId": msg.WebsiteID,
				"eventType": string(eventType),
				"status":    string(NotifSkipped),
				"reason":    "channel_not_registered",
			})
			continue
		}

		// 6. Resolve user email once
		if userEmail == "" {
			userEmail, err = nw.websiteRepo.GetUserEmailByWebsiteID(
				ctx,
				msg.WebsiteID,
			)
			if err != nil {
				return err
			}
		}

		// 7. Send notification
		providerMessageID, sendErr := channel.Send(ctx, channels.NotificationPayload{
			WebsiteID:  msg.WebsiteID,
			EventType:  eventType,
			OccurredAt: msg.OccurredAt,
			Target:     userEmail,
		})

		var providerMessageIDPtr *string
		if providerMessageID != "" {
			providerMessageIDPtr = &providerMessageID
		}

		deliveryStatus := "sent"
		if sendErr != nil {
			deliveryStatus = "failed"
		}

		// 8. Persist notification attempt
		err = nw.notificationLogService.RecordNotificationAttempt(
			ctx,
			service.RecordNotificationAttemptInput{
				WebsiteID:         msg.WebsiteID,
				IncidentID:        incidentID,
				RegionID:          msg.RegionID,
				Channel:           rule.Channel,
				Recipient:         userEmail,
				PrevStatus:        msg.PrevStatus,
				CurrentStatus:     msg.CurrentStatus,
				DeliveryStatus:    deliveryStatus,
				ProviderMessageID: providerMessageIDPtr,
				SentAt:            time.Now().UTC(),
			},
		)
		if err != nil {
			return err
		}

		// 9. Handle send result
		if sendErr != nil {
			LogNotification(map[string]string{
				"websiteId": msg.WebsiteID,
				"eventType": string(eventType),
				"channel":   rule.Channel,
				"status":    "failed",
			})
			continue
		}

		sentAtLeastOnce = true

		LogNotification(map[string]string{
			"websiteId": msg.WebsiteID,
			"eventType": string(eventType),
			"channel":   rule.Channel,
			"status":    string(NotifSent),
		})
	}

	if !sentAtLeastOnce {
		LogNotification(map[string]string{
			"websiteId": msg.WebsiteID,
			"eventType": string(eventType),
			"status":    string(NotifNoOp),
		})
	}

	return nil
}
